### PR TITLE
Add compaction filter factory API

### PR DESCRIPTION
**Pull request by unrealhoang**
And other option configuration.

### CODE REVIEW

File: src/compaction_filter_factory.rs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    ffi::rocksdb_compactionfilter_create(
+        filter_ptr as *mut c_void,
+        Some(compaction_filter::destructor_callback::<F::Filter>),
+        Some(compaction_filter::filter_callback::<F::Filter>),
+        Some(compaction_filter::name_callback::<F::Filter>),
+    )
+}
+
+#[cfg(test)]
+mod test {
```
Comment (DarkEld3r) [inline src/compaction_filter_factory.rs:76]: Please rename the module to `tests` - it should fix the Clippy error.

File: src/compaction_filter.rs [hunk 2]
```diff
@@ -51,19 +56,32 @@ where
...
+    fn filter(&mut self, level: u32, key: &[u8], value: &[u8]) -> Decision {
+        (self.filter_fn)(level, key, value)
+    }
+}
+
+pub unsafe extern "C" fn destructor_callback<F>(raw_cb: *mut c_void)
+where
+    F: CompactionFilter,
+{
+    let _: Box<F> = mem::transmute(raw_cb as *mut F);
```
[the remark this answers is not in the document]
Comment (unrealhoang) [reply src/compaction_filter.rs:76]: This is the old code, I'm not really sure about the intention so I kept it the same. But I can change as you suggested.

[the remark this answers is not in the document]
Comment (DarkEld3r) [reply src/compaction_filter.rs:76]: I think it is better to use `Box::from_raw` in both places.

**Comment by DarkEld3r** [comment]
The documentation suggests that `wildcard_imports` lint shouldn't be triggered in any module which name contains "test". I have no clues why it isn't working in our case. I have submitted the [issue](https://github.com/rust-lang/rust-clippy/issues/5764), lets hope someone will suggest a workaround.

**Comment by DarkEld3r** [comment]
The issue with the Clippy is fixed, but the release should be available next week, so I guess it is OK to suppress this warning for now.

File: src/compaction_filter.rs [hunk 3]
```diff
@@ -32,6 +31,11 @@ pub enum Decision {
     Change(&'static [u8]),
 }
 
+pub trait CompactionFilter {
```
Comment (DarkEld3r) [inline src/compaction_filter.rs:34]: It would be nice to have documentation for this trait.

File: src/compaction_filter_factory.rs [hunk 4]
```diff
@@ -0,0 +1,129 @@
+use std::ffi::CStr;
+
+use libc::{self, c_char, c_void};
+
+use crate::{
+    compaction_filter::{self, CompactionFilter},
+    ffi,
+};
+
+pub trait CompactionFilterFactory {
```
Comment (DarkEld3r) [inline src/compaction_filter_factory.rs:10]: Same here. :upside_down_face:

**Comment by unrealhoang** [comment]
@DarkEld3r Thank you for your review, can you take a another look?

**Pull request merged**